### PR TITLE
*:add opt_rule_blacklist in mysql tables.

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
In some product environments. We found that a switch to disable some logical optimizing rules is convenient for troubleshooting.

### What is changed and how it works?
Like `expr_pushdown_blacklist`, just add a `opt_rule_blacklist`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 
Side effects

 - None

Related changes

 - Need to update the documentation

### CODE REVIEW

File: executor/disabled_optimize_list.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	sql := "select HIGH_PRIORITY name, type from mysql.disabled_optimize_list"
+	rows, _, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, sql)
+	if err != nil {
+		return err
+	}
+	newExprPushdownBlacklist := make(map[string]struct{})
+	newDisabledLogicalRules := make(map[string]struct{})
+	for _, row := range rows {
+		name := row.GetString(0)
+		if row.GetString(1) == "expr_push_down" {
```
Comment (XuHuaiyu) [inline executor/disabled_optimize_list.go]: Maybe we should report a warning if row.GetString(1) is not "expr_push_down" neither "logical_rule".

Comment (lzmhhh123) [reply executor/disabled_optimize_list.go]: Okay.

File: executor/disabled_optimize_list.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	baseExecutor
+}
+
+// Next implements the Executor Next interface.
+func (e *ReloadDisabledOptimizeListExec) Next(ctx context.Context, _ *chunk.Chunk) error {
+	return LoadDisabledOptimizeList(e.ctx)
+}
+
+// LoadDisabledOptimizeList loads the latest data from table mysql.disabled_optimize_list.
+func LoadDisabledOptimizeList(ctx sessionctx.Context) (err error) {
```
Comment (zz-jason) [inline executor/disabled_optimize_list.go]: As we discussed in the paser pr, load disabled rule list should not load the disabled expr list anymore.

File: executor/disabled_optimize_list_test.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/pingcap/tidb/util/testkit"
+)
+
+func (s *testSuite2) TestReloadDisabledOptimizeList(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("create database disabled_optimize_list")
+	tk.MustExec("use disabled_optimize_list")
+	tk.MustExec("create table t (a int)")
+	// test disable expr push down.
+	tk.MustQuery("desc select * from t where a < 1").Check(testkit.Rows(
```
Comment (zz-jason) [inline executor/disabled_optimize_list_test.go]: it's better to add these tests to `cmd/explaintest/`

File: session/bootstrap.go [hunk 4]
```diff
@@ -264,6 +264,12 @@ const (
 	CreateExprPushdownBlacklist = `CREATE TABLE IF NOT EXISTS mysql.expr_pushdown_blacklist (
 		name char(100) NOT NULL
 	);`
+
+	// CreateOptRuleBlacklist stores the list of disabled optimizing operations.
+	CreateOptRuleBlacklist = `CREATE TABLE IF NOT EXISTS mysql.opt_rule_blacklist (
+		name char(100) NOT NULL,
+		type char(20) check(type="logical_rule") NOT NULL
```
Comment (zz-jason) [inline session/bootstrap.go]: Do we still need the `type` field if we separate the blacklist of rules and pushdown functions?

Comment (lzmhhh123) [reply session/bootstrap.go]: Yes, in the future, after the cascade planner complete, this field can also be used for it.

File: executor/opt_rule_blacklist.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	if err != nil {
+		return err
+	}
+	newDisabledLogicalRules := set.NewStringSet()
+	for _, row := range rows {
+		name := strings.ToLower(row.GetString(0))
+		if row.GetString(1) == "logical_rule" {
+			newDisabledLogicalRules.Insert(name)
+		} else {
+			ctx.GetSessionVars().StmtCtx.AppendWarning(errors.New("type field of mysql.opt_rule_blacklist can only be \"expr_push_down\" or \"logical_rule\""))
```
Comment (lamxTyler) [inline executor/opt_rule_blacklist.go]: Outdated now?

Comment (lzmhhh123) [reply executor/opt_rule_blacklist.go]: removed.

File: session/bootstrap.go (hunk 4 above)
Comment (lamxTyler) [inline session/bootstrap.go]: Why add check` since tidb does not support it?

Comment (zz-jason) [reply session/bootstrap.go]: Do you mean to use this field to indicates which planner this rule belonging to?

File: go.mod [hunk 6]
```diff
@@ -68,11 +68,12 @@ require (
...
-	golang.org/x/time v0.0.0-20190308202827-9d24e82272b4 // indirect
 	golang.org/x/tools v0.0.0-20190130214255-bb1329dc71a0
 	google.golang.org/genproto v0.0.0-20190108161440-ae2f86662275 // indirect
 	google.golang.org/grpc v1.17.0
 	gopkg.in/natefinch/lumberjack.v2 v2.0.0
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20180531100431-4c381bd170b4
 	sourcegraph.com/sourcegraph/appdash-data v0.0.0-20151005221446-73f23eafcf67
 )
+
+replace github.com/pingcap/parser => github.com/pingcap/parser v0.0.0-20190716105515-ded7e0436a50
```
Comment (lamxTyler) [inline go.mod]: Seems that the parser is already merged.

File: session/bootstrap.go (hunk 4 above)
Comment (lzmhhh123) [reply session/bootstrap.go]: It's just a reminder for people. Like code comments, or `show create table mysql.opt_rule_blacklist` will contains it.

Comment (lzmhhh123) [reply session/bootstrap.go]: Yes.

File: executor/reload_expr_pushdown_blacklist.go [hunk 7]
```diff
@@ -42,8 +42,8 @@ func LoadExprPushdownBlacklist(ctx sessionctx.Context) (err error) {
 	}
 	newBlacklist := make(map[string]struct{})
 	for _, row := range rows {
-		name := row.GetString(0)
-		newBlacklist[strings.ToLower(name)] = struct{}{}
+		name := strings.ToLower(row.GetString(0))
```
Comment (zz-jason) [inline executor/reload_expr_pushdown_blacklist.go]: can we avoid the function call to `strings.ToLower()`?

File: session/bootstrap.go (hunk 4 above)
Comment (zz-jason) [reply session/bootstrap.go]: I think we don't need this field even when we want to disable some rules in the cascades planner. The rule name in cascades planner is different from the current planner. BTW, we are going to replace the current planner with the cascades planner in the near future. At that time, the only valid type is `cascades planner`. How about only use a `name` field?

**Comment by zz-jason** [comment]
BTW, we need to document all the rules which can be disabled for users.

File: session/bootstrap.go (hunk 4 above)
Comment (lzmhhh123) [reply session/bootstrap.go]: OK

**Pull request merged**